[docs] LLM inference (#29791)
huggingface/transformers: docs/source/en/llm_optims.md

## Static kv-cache and torch.compile

During decoding, a LLM computes the key-value (kv) values for each input token and since it is autoregressive, it computes the same kv values each time since the generated output becomes part of the input now. This is not very efficient because you're recomputing the same kv values each time.

To optimize this, you can use a kv-cache to store the past values instead of recomputing them each time. However, since the kv-cache grows with each generation step and is dynamic, it prevents you from taking advantage of [torch.compile](./perf_torch_compile), a powerful optimization tool that fuses Python code into fast and optimized kernels.

The *static kv-cache* solves this issue by fixing the kv-cache size to a maximum value which allows you to combine it with torch.compile for up to a 4x speed up.

> [!WARNING]

> > Currently, only [Gemma](./model_doc/gemma) and [Llama](./model_doc/llama2) models support static kv-cache and torch.compile.

Reviewer: and cohere I think

For this example, let's load the [Gemma](https://hf.co/google/gemma-2b) model.

```py
from transformers import AutoTokenizer, AutoModelForCausalLM

tokenizer = AutoTokenizer.from_pretrained("google/gemma-2b")

model = AutoModelForCausalLM.from_pretrained(
    "google/gemma-2b", device_map="auto"
)
```

There are two ways you can configure the model to use a static kv-cache.

<hfoptions id="static-kv">
<hfoption id="generation_config">

Access the model's `generation_config` attribute and set the `cache_implementation` to "static".

```py
model.generation_config.cache_implementation = "static"
```

Call torch.compile on the model to compile it with the static kv-cache.

```py
compiled_model = torch.compile(model, mode="reduce-overhead", fullgraph=True)
input_text = "Planck's constant is"
input_ids = tokenizer(input_text, return_tensors="pt").to("cuda")

> outputs = compiled_model.generate(**input_ids)

Reviewer: generate is not yet compatible with torch compile so would only recommend the forward pass to be shown here! And talk about 4x!

```

</hfoption>
<hfoption id="setup_cache">

> Access the model's `_set_cache` method and pass it the [`StaticCache`] class. This is a more flexible method because it allows you to configure parameters like the maximum batch size and sequence length.

Reviewer 2: I think it would be great to think about changing the method name from `_set_cache` to `set_cache` if it is not supposed to be internal.
Also, I think it would be good to explain what is the value that should be put in the `max_cache_len` (maximum number of tokens if I got it right).

```py
model._setup_cache(StaticCache, max_batch_size=1, max_cache_len=128)
```

Call torch.compile on the model to compile it with the static kv-cache.

```py
compiled_model = torch.compile(model, mode="reduce-overhead", fullgraph=True)
input_text = "Planck's constant is"
input_ids = tokenizer(input_text, return_tensors="pt").to("cuda")
outputs = compiled_model(input_ids)
```

</hfoption>
</hfoptions>

> ## Speculative decoding

Reviewer 3: Shameless plug: https://huggingface.co/blog/assisted-generation 😉

```py
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

device = "cuda" if torch.cuda.is_available() else "cpu"

tokenizer = AutoTokenizer.from_pretrained("facebook/opt-1.3b")
inputs = tokenizer("The second law of thermodynamics states", return_tensors="pt").to(device)

model = AutoModelForCausalLM.from_pretrained("facebook/opt-1.3b").to(device)
outputs = model.generate(**inputs, prompt_lookup_num_tokens=3, do_sample=True, temperature=0.7)
print(tokenizer.batch_decode(outputs, skip_special_tokens=True))
["The second law of thermodynamics states that energy cannot be created nor destroyed. It's not a"]
```

</hfoption>
</hfoptions>

## Attention optimizations

A known issue with transformer models is that the self-attention mechanism grows quadratically in compute and memory with the number of input tokens. This limitation is only magnified in LLMs which can handle much longer sequences. To address this, try FlashAttention2 or PyTorch's scaled dot product attention (SDPA), which are more memory efficient implementations and can accelerate inference.

### FlashAttention-2

> FlashAttention and [FlashAttention-2](./perf_infer_gpu_one#flashattention-2) breaks up the attention computation into smaller chunks and reduces the number of intermediate read/write operations to GPU memory to speed up inference. FlashAttention-2 improves on the original FlashAttention algorithm by also parallelizing over sequence length dimension and better partitioning work on the hardware to reduce synchronization and communication overhead.

Reviewer 4: ```suggestion
FlashAttention and [FlashAttention-2](./perf_infer_gpu_one#flashattention-2) break up the attention computation into smaller chunks and reduce the number of intermediate read/write operations to GPU memory to speed up inference. FlashAttention-2 improves on the original FlashAttention algorithm by also parallelizing over sequence length dimension and better partitioning work on the hardware to reduce synchronization and communication overhead.
```

To use FlashAttention-2, set `attn_implementation="flash_attention_2"` in the [`~PreTrainedModel.from_pretrained`] method.

```py
from transformers import AutoModelForCausalLM

model = AutoModelForCausalLM.from_pretrained(
    "google/gemma-2b",
    torch_dtype=bfloat16,
    attn_implementation="flash_attention_2",
    load_in_8bit=True,
)
```

### PyTorch scaled dot product attention

> Scaled dot product attention (SDPA) is automatically enabled in PyTorch 2.0 and it supports FlashAttention, xFormers, and a C++ PyTorch implementation. SDPA chooses the most performant attention algorithm based on your hardware, but you could also explicitly enable or disable the attention algorithm you want to use.

Reviewer 2: You say "SDPA chooses the most performan algorithm based on your hardware", but that is true for CUDA, not other backends (mps, xla, etc), right?

Author: Correct, I'll update and clarify this. Thanks :)

> [!TIP]
> There are many quantization libraries (see the [Quantization](./quantization) guide for more details) available, such as Quanto, AQLM, AWQ, and AutoGPTQ. Feel free to try them out and see which one works best for your use case. We also recommend reading the [Overview of natively supported quantization schemes in 🤗 Transformers](https://hf.co/blog/overview-quantization-transformers) blog post which compares AutoGPTQ and bitsandbytes.

Use the Model Memory Calculator below to estimate and compare how much memory is required to load a model. For example, try estimating how much memory it costs to load [Mistral-7B-v0.1](https://huggingface.co/mistralai/Mistral-7B-v0.1).

<iframe
	src="https://hf-accelerate-model-memory-usage.hf.space"
	frameborder="0"
	width="850"
	height="450"
></iframe>

To load Mistral-7B-v0.1 in half-precision, set the `torch_dtype` parameter in the [`~transformers.AutoModelForCausalLM.from_pretrained`] method to `torch.bfloat16`. This only costs 13.74GB of memory.

```py
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

tokenizer = AutoTokenizer.from_pretrained("mistralai/Mistral-7B-v0.1")

> model = AutoModelForCausalLM.from_pretrained(
>     "mistralai/Mistral-7B-v0.1", torch_dtype=torch.bfloat16, device_map="auto",
> ).to("cuda")

Reviewer 5: same issue
```suggestion
model = AutoModelForCausalLM.from_pretrained(
    "mistralai/Mistral-7B-v0.1", torch_dtype=torch.bfloat16, device_map="auto",
)
```

```

To load a quantized model (8-bit or 4-bit) for inference, try [bitsandbytes](https://hf.co/docs/bitsandbytes) and set the `load_in_4bit` or `load_in_8bit` parameters to `True`. Loading the model in 8-bits only costs 6.87 GB of memory.

> 

Reviewer 5: Since we are deprecating `load_in_4bit` and `load_in_8bit`, I suggest using `quantization_config` instead and pass a `BitsAndBytesConfig` object.
If the user wants to do 8-bit quantization, he needs to pass `BitsAndBytesConfig(load_in_8bit=True)`. For 4-bit quantization, he needs to pass `BitsAndBytesConfig(load_in_4bit=True)`.

```py
from transformers import AutoTokenizer, AutoModelForCausalLM
import torch

tokenizer = AutoTokenizer.from_pretrained("mistralai/Mistral-7B-v0.1")

> model = AutoModelForCausalLM.from_pretrained(
>     "mistralai/Mistral-7B-v0.1", device_map="auto", load_in_8bit=True,
> ).to("cuda")

Reviewer 5: We shouldn't move the model when we use device_map="auto" since the model is already dispatch the GPUs. It might trigger some issues in the multi-gpu case otherwise. 
```suggestion
model = AutoModelForCausalLM.from_pretrained(
    "mistralai/Mistral-7B-v0.1", device_map="auto", load_in_8bit=True,
)
```

The `_setup_cache` method doesn't support [`GenerationMixin.generate`] yet, so this method is a bit more involved. You'll need to write your own function to decode the next token given the current token and position and cache position of previously generated tokens.

```py
from transformers import LlamaTokenizer, LlamaForCausalLM, StaticCache, logging
from transformers.testing_utils import CaptureLogger
import torch

prompts = [
    "Simply put, the theory of relativity states that ",
    "My favorite all time favorite condiment is ketchup.",
]

NUM_TOKENS_TO_GENERATE = 40
torch_device = "cuda"

tokenizer = LlamaTokenizer.from_pretrained("meta-llama/Llama-2-7b-hf", pad_token="</s>", padding_side="right")
model = LlamaForCausalLM.from_pretrained("meta-llama/Llama-2-7b-hf", device_map="sequential")
inputs = tokenizer(prompts, return_tensors="pt", padding=True).to(model.device)

def decode_one_tokens(model, cur_token, input_pos, cache_position):
    logits = model(
        cur_token, position_ids=input_pos, cache_position=cache_position, return_dict=False, use_cache=True
    )[0]
    new_token = torch.argmax(logits[:, -1], dim=-1)[:, None]
    return new_token
```

> The are two important things you must do to enable static kv-cache and torch.compile with the `_setup_cache` method:

Reviewer 6: I'm not sure about advertising using private methods - this means they're effectively public and need to have the same strict backwards compatibility handling of public methods

Author: I think it's ok with a warning that this is still experimental and its design will probably change in the future, so the user should be aware that some things may not work as expected, but in the meantime, you can still get excited/try this out? 

Would it help if I added a sentence about how it may not be backward compatible?

Reviewer 6: That works for me!

1. Access the model's `_setup_cache` method and pass it the [`StaticCache`] class. This is a more flexible method because it allows you to configure parameters like the maximum batch size and sequence length.

2. Call torch.compile on the model to compile the forward pass with the static kv-cache.

```py
batch_size, seq_length = inputs["input_ids"].shape
with torch.no_grad():
     model._setup_cache(StaticCache, 2, max_cache_len=4096)
     cache_position = torch.arange(seq_length, device=torch_device)
     generated_ids = torch.zeros(
         batch_size, seq_length + NUM_TOKENS_TO_GENERATE + 1, dtype=torch.int, device=torch_device
     )
     generated_ids[:, cache_position] = inputs["input_ids"].to(torch_device).to(torch.int)

     logits = model(**inputs, cache_position=cache_position, return_dict=False, use_cache=True)[0]
     next_token = torch.argmax(logits[:, -1], dim=-1)[:, None]
     generated_ids[:, seq_length] = next_token[:, 0]

     decode_one_tokens = torch.compile(decode_one_tokens, mode="reduce-overhead", fullgraph=True)
     cache_position = torch.tensor([seq_length + 1], device=torch_device)
     for _ in range(1, NUM_TOKENS_TO_GENERATE):

>          with torch.backends.cuda.sdp_kernel(enable_flash=False, enable_mem_efficient=False, enable_math=True):

Reviewer 6: It would be good to have a comment here explaining the value of this context manager

             next_token = decode_one_tokens(model, next_token.clone(), None, cache_position)
             generated_ids[:, cache_position] = next_token.int()
         cache_position += 1

text = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)
text
['Simply put, the theory of relativity states that 1) the speed of light is constant, 2) the speed of light is the same for all observers, and 3) the laws of physics are the same for all observers.',
 'My favorite all time favorite condiment is ketchup. I love it on everything. I love it on my eggs, my fries, my chicken, my burgers, my hot dogs, my sandwiches, my salads, my p']

> ```

Reviewer 6: For both of the compile options - I think it would be really compelling for a user if they had actual numbers e.g. fdo we get a 4x speed up under all cases? Do we get the same speed-up for these two cache approaches?

Author: Totally agree, I'll ask @gante about this 🙂 

In the meantime, would it be ok to merge this PR first and add this in a separate PR or would you prefer to wait?

Reviewer 6: Sure!

Reviewer: Sure

Basic inference is slow because LLMs have to be called repeatedly to generate the next token. The input sequence increases as generation progresses, which takes longer and longer for the LLM to process. LLMs also have billions of parameters, making it a challenge to store and handle all those weights in memory.

This guide will show you how to use the optimization techniques available in Transformers to accelerate LLM inference.

> [!TIP]
> Hugging Face also provides [Text Generation Inference (TGI)](https://hf.co/docs/text-generation-inference), a library dedicated to deploying and serving highly optimized LLMs for inference. It includes more optimization features not included in Transformers, such as continuous batching for increasing throughput and tensor parallelism for multi-GPU inference.

## Static kv-cache and torch.compile

During decoding, a LLM computes the key-value (kv) values for each input token and since it is autoregressive, it computes the same kv values each time because the generated output becomes part of the input now. This is not very efficient because you're recomputing the same kv values each time.

> To optimize this, you can use a kv-cache to store the past values instead of recomputing them each time. However, since the kv-cache grows with each generation step and is dynamic, it prevents you from taking advantage of [torch.compile](./perf_torch_compile), a powerful optimization tool that fuses PyTorch code into fast and optimized kernels.

Reviewer: ```suggestion
To optimize this, you can use a kv-cache to store the past keys and values instead of recomputing them each time. However, since the kv-cache grows with each generation step and is dynamic, it prevents you from taking advantage of [torch.compile](./perf_torch_compile), a powerful optimization tool that fuses PyTorch code into fast and optimized kernels.
```

> The *static kv-cache* solves this issue by fixing the kv-cache size to a maximum value which allows you to combine it with torch.compile for up to a 4x speed up.

Reviewer: ```suggestion
The *static kv-cache* solves this issue by pre-allocating the kv-cache size to a maximum value which allows you to combine it with torch.compile for up to a 4x speed up.
```
